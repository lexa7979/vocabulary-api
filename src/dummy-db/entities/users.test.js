const { getTestId, _testHelpers: TestIdHelpers } = require('../testIds');

const Users = require('./users');

const { ASYNC, bold, EXPECTS, IS_ACCESSIBLE, REJECTS, RESOLVES } = require('../../../test');

const Global = {};

describe(`Test-data service "users"`, () => {
  runTestsAboutGetUser();
  runTestsAboutListAllUsers();
  runTestsAboutGetActiveWord();
  runTestsAboutListAllActiveWordsOfUser();
});

function runTestsAboutGetUser() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold('getUser()')} which`, () => {
    const { getUser } = Users;

    it(IS_ACCESSIBLE, () => expect(getUser).toBeFunction());

    it(`${EXPECTS} one argument (id)`, () => expect(getUser).toHaveLength(1));

    it(`- when used with valid user-ID - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const result = await getUser(getTestId('user1'));

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result);
      expect(copy).toMatchSnapshot(`> user1 <`);
    });

    it(`- when used with an invalid user-ID - ${REJECTS} as expected`, async () => {
      await expect(() => getUser(getTestId('invalid'))).rejects.toThrow('invalid ID');
    });
  });
}

function runTestsAboutListAllUsers() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold('listAllUsers()')} which`, () => {
    const { listAllUsers } = Users;

    it(IS_ACCESSIBLE, () => expect(listAllUsers).toBeFunction());

    it(`${EXPECTS} no arguments`, () => expect(listAllUsers).toHaveLength(0));

    it(`- when used w/o arguments - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const result = await listAllUsers();

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result);
      expect(copy).toMatchSnapshot(`> all users <`);
    });

    it(`- when used with argument "first" - ${RESOLVES} as expected`, async () => {
      const result = await listAllUsers(1, undefined);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result).map(item => item.id);
      expect(copy).toMatchInlineSnapshot(`
        Array [
          "(ID:user1)",
        ]
      `);
    });

    it(`- when used with argument "offset" - ${RESOLVES} as expected`, async () => {
      const result = await listAllUsers(undefined, 1);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result).map(item => item.id);
      expect(copy).toMatchInlineSnapshot(`
        Array [
          "(ID:user2)",
        ]
      `);
    });
  });
}

function runTestsAboutGetActiveWord() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold('getActiveWord()')} which`, () => {
    const { getActiveWord } = Users;

    it(IS_ACCESSIBLE, () => expect(getActiveWord).toBeFunction());

    it(`${EXPECTS} one argument (activeWordId)`, () => expect(getActiveWord).toHaveLength(1));

    it(`- when used with valid ID - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const result = await getActiveWord(getTestId('user1-activeWord1'));

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result);
      expect(copy).toMatchSnapshot(`> user1-activeWord1 <`);
    });

    it(`- when used with invalid ID - ${REJECTS} as expected`, async () => {
      await expect(() => getActiveWord(getTestId('invalid'))).rejects.toThrow('invalid ID');
    });
  });
}

function runTestsAboutListAllActiveWordsOfUser() {
  describe(`${_nextListHint()} has an ${ASYNC} function ${bold(
    'listAllActiveWordsOfUser()'
  )} which`, () => {
    const { listAllActiveWordsOfUser } = Users;

    it(IS_ACCESSIBLE, () => expect(listAllActiveWordsOfUser).toBeFunction());

    it(`${EXPECTS} one argument (userId)`, () => expect(listAllActiveWordsOfUser).toHaveLength(1));

    it(`- when used w/o arguments - ${RESOLVES} with empty array`, async () => {
      const result = await listAllActiveWordsOfUser();
      expect(result).toEqual([]);
    });

    it(`- when used with valid user-ID - ${RESOLVES} as expected ${_nextSnapshotHint()}`, async () => {
      const result1 = await listAllActiveWordsOfUser(getTestId('user1'));
      expect(result1).toBeArrayOfSize(2);

      const result2 = await listAllActiveWordsOfUser(getTestId('user2'));
      expect(result2).toBeArrayOfSize(0);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result1);
      expect(copy).toMatchSnapshot(`> all active-words of user1 <`);
    });

    it(`- when used with valid user-ID and argument "first" - ${RESOLVES} as expected`, async () => {
      const result = await listAllActiveWordsOfUser(getTestId('user1'), 1, undefined);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result).map(item => item.id);
      expect(copy).toMatchInlineSnapshot(`
        Array [
          "(ID:user1-activeWord1)",
        ]
      `);
    });

    it(`- when used with valid user-ID and argument "offset" - ${RESOLVES} as expected`, async () => {
      const result = await listAllActiveWordsOfUser(getTestId('user1'), undefined, 1);

      const copy = TestIdHelpers.copyObjectButRemoveUUIDs(result).map(item => item.id);
      expect(copy).toMatchInlineSnapshot(`
        Array [
          "(ID:user1-activeWord2)",
        ]
      `);
    });
  });
}

function _nextListHint() {
  Global.nextListItemId = Global.nextListItemId == null ? 1 : Global.nextListItemId + 1;
  return `(${Global.nextListItemId})`;
}

function _nextSnapshotHint() {
  Global.nextSnapshotId = Global.nextSnapshotId == null ? 1 : Global.nextSnapshotId + 1;
  return `[-> check snapshot ${Global.nextSnapshotId}]`;
}
