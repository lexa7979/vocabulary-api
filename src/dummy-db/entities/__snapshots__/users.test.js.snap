// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test-data service "users" (1) has an [1;93masync[0;90m function [1;93mgetUser()[0;90m which - when used with valid ID - [0;92mresolves[0;90m as expected [-> check snapshot 1]: > user1 < 1`] = `
{
  "currStep": 8,
  "currentLessons": [
    {
      "doneAt": "(TS:-2min)",
      "flectionId": "(ID:adjective-flection3)",
      "result": "correct",
      "wordId": "(ID:word2)",
    },
    {
      "flectionId": "(ID:verb-flection1)",
      "wordId": "(ID:word1)",
    },
  ],
  "currentSession": {
    "startedAt": "(TS:-10min)",
  },
  "id": "(ID:user1)",
  "learnPath": [
    {
      "flectionId": "(ID:adjective-flection2)",
      "result": 100,
      "step": 7,
      "wordId": "(ID:word3)",
    },
    {
      "flectionId": "(ID:verb-flection4)",
      "result": null,
      "step": 8,
      "wordId": "(ID:word1)",
    },
  ],
  "login": "user1",
  "name": "User 1",
}
`;

exports[`Test-data service "users" (2) has an [1;93masync[0;90m function [1;93mlistAllUsers()[0;90m which - when used w/o arguments - [0;92mresolves[0;90m as expected [-> check snapshot 2]: > all users < 1`] = `
[
  {
    "currStep": 8,
    "currentLessons": [
      {
        "doneAt": "(TS:-2min)",
        "flectionId": "(ID:adjective-flection3)",
        "result": "correct",
        "wordId": "(ID:word2)",
      },
      {
        "flectionId": "(ID:verb-flection1)",
        "wordId": "(ID:word1)",
      },
    ],
    "currentSession": {
      "startedAt": "(TS:-10min)",
    },
    "id": "(ID:user1)",
    "learnPath": [
      {
        "flectionId": "(ID:adjective-flection2)",
        "result": 100,
        "step": 7,
        "wordId": "(ID:word3)",
      },
      {
        "flectionId": "(ID:verb-flection4)",
        "result": null,
        "step": 8,
        "wordId": "(ID:word1)",
      },
    ],
    "login": "user1",
    "name": "User 1",
  },
  {
    "currStep": 0,
    "id": "(ID:user2)",
    "learnPath": [],
    "login": "user2",
    "name": "User 2",
  },
]
`;

exports[`Test-data service "users" (3) has an [1;93masync[0;90m function [1;93mgetActiveWord()[0;90m which - when used with valid ID - [0;92mresolves[0;90m as expected [-> check snapshot 3]: > 1st activeWord of user1 < 1`] = `
{
  "events": [
    {
      "timestamp": "(TS:-5d)",
      "type": "add",
    },
    {
      "flectionId": "(ID:verb-flection3)",
      "result": "wrong",
      "timestamp": "(TS:-2d)",
      "type": "flection-result",
    },
    {
      "flectionId": "(ID:verb-flection1)",
      "result": "correct",
      "timestamp": "(TS:yesterday)",
      "type": "flection-result",
    },
  ],
  "id": "(ID:user1-activeWord1)",
  "learnProgress": [
    {
      "changedAt": null,
      "currGroup": 0,
      "flectionId": "(ID:verb-flection1)",
    },
    {
      "changedAt": null,
      "currGroup": 0,
      "flectionId": "(ID:verb-flection2)",
    },
    {
      "changedAt": null,
      "currGroup": 0,
      "flectionId": "(ID:verb-flection3)",
    },
    {
      "changedAt": null,
      "currGroup": 0,
      "flectionId": "(ID:verb-flection4)",
    },
    {
      "changedAt": null,
      "currGroup": 0,
      "flectionId": "(ID:verb-flection5)",
    },
  ],
  "nextLessons": [
    {
      "currGroup": 1,
      "earliestAt": "(TS:tomorrow)",
      "flectionId": "(ID:verb-flection1)",
    },
    {
      "currGroup": 0,
      "earliestAt": "(TS:-5d)",
      "flectionId": "(ID:verb-flection2)",
    },
    {
      "currGroup": 0,
      "earliestAt": "(TS:yesterday)",
      "flectionId": "(ID:verb-flection3)",
    },
    {
      "currGroup": 0,
      "earliestAt": "(TS:-5d)",
      "flectionId": "(ID:verb-flection4)",
    },
    {
      "currGroup": 0,
      "earliestAt": "(TS:-5d)",
      "flectionId": "(ID:verb-flection5)",
    },
  ],
  "userId": "(ID:user1)",
  "wordId": "(ID:word1)",
}
`;

exports[`Test-data service "users" (4) has an [1;93masync[0;90m function [1;93mlistAllActiveWordsOfUser()[0;90m which - when used with valid ID - [0;92mresolves[0;90m as expected [-> check snapshot 4]: > all active-words of user1 < 1`] = `
[
  {
    "events": [
      {
        "timestamp": "(TS:-5d)",
        "type": "add",
      },
      {
        "flectionId": "(ID:verb-flection3)",
        "result": "wrong",
        "timestamp": "(TS:-2d)",
        "type": "flection-result",
      },
      {
        "flectionId": "(ID:verb-flection1)",
        "result": "correct",
        "timestamp": "(TS:yesterday)",
        "type": "flection-result",
      },
    ],
    "id": "(ID:user1-activeWord1)",
    "learnProgress": [
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:verb-flection1)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:verb-flection2)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:verb-flection3)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:verb-flection4)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:verb-flection5)",
      },
    ],
    "nextLessons": [
      {
        "currGroup": 1,
        "earliestAt": "(TS:tomorrow)",
        "flectionId": "(ID:verb-flection1)",
      },
      {
        "currGroup": 0,
        "earliestAt": "(TS:-5d)",
        "flectionId": "(ID:verb-flection2)",
      },
      {
        "currGroup": 0,
        "earliestAt": "(TS:yesterday)",
        "flectionId": "(ID:verb-flection3)",
      },
      {
        "currGroup": 0,
        "earliestAt": "(TS:-5d)",
        "flectionId": "(ID:verb-flection4)",
      },
      {
        "currGroup": 0,
        "earliestAt": "(TS:-5d)",
        "flectionId": "(ID:verb-flection5)",
      },
    ],
    "userId": "(ID:user1)",
    "wordId": "(ID:word1)",
  },
  {
    "events": [
      {
        "timestamp": "(TS:-3d)",
        "type": "add",
      },
      {
        "flectionId": "(ID:adjective-flection5)",
        "result": "correct",
        "timestamp": "(TS:-2d)",
        "type": "flection-result",
      },
      {
        "flectionId": "(ID:adjective-flection2)",
        "result": "wrong",
        "timestamp": "(TS:yesterday)",
        "type": "flection-result",
      },
      {
        "flectionId": "(ID:adjective-flection3)",
        "result": "correct",
        "timestamp": "(TS:-2min)",
        "type": "flection-result",
      },
    ],
    "id": "(ID:user1-activeWord2)",
    "learnProgress": [
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:adjective-flection1)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:adjective-flection2)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:adjective-flection3)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:adjective-flection4)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:adjective-flection5)",
      },
      {
        "changedAt": null,
        "currGroup": 0,
        "flectionId": "(ID:adjective-flection6)",
      },
    ],
    "nextLessons": [
      {
        "currGroup": 0,
        "earliestAt": "(TS:-3d)",
        "flectionId": "(ID:adjective-flection1)",
      },
      {
        "currGroup": 0,
        "earliestAt": "(TS:now)",
        "flectionId": "(ID:adjective-flection2)",
      },
      {
        "currGroup": 1,
        "earliestAt": "(TS:+1d 23h 58min)",
        "flectionId": "(ID:adjective-flection3)",
      },
      {
        "currGroup": 0,
        "earliestAt": "(TS:-3d)",
        "flectionId": "(ID:adjective-flection4)",
      },
      {
        "currGroup": 1,
        "earliestAt": "(TS:now)",
        "flectionId": "(ID:adjective-flection5)",
      },
      {
        "currGroup": 0,
        "earliestAt": "(TS:-3d)",
        "flectionId": "(ID:adjective-flection6)",
      },
    ],
    "userId": "(ID:user1)",
    "wordId": "(ID:word3)",
  },
]
`;
